tls/extensions/s2n_client_psk.c

```diff
@@ -159,49 +159,47 @@ static S2N_RESULT s2n_match_psk_identity(struct s2n_array *known_psks, const str
     return S2N_RESULT_OK;
 }
 
-static S2N_RESULT s2n_select_psk_identity(struct s2n_connection *conn, struct s2n_psk_identity *identities, size_t identities_length)
+/* Find the first of the server's known PSK identities that matches the client's offered wire identities.
+ * This method compares all local identities to all wire identities.
+ *
+ * While both the client's offered identities and whether a match was found are public, we should make an attempt
+ * to keep the server's known identities a secret. We will make comparisons to the server's identities constant
+ * time (to hide partial matches) and not end the search early when a match is found (to hide the ordering).
+ *
+ * Keeping these comparisons constant time is not high priority. There's no known attack using these timings,
+ * and an attacker could probably guess the server's known identities just by observing the public identities
+ * sent by clients.
+ */
+static S2N_RESULT s2n_select_psk_identity(struct s2n_connection *conn, struct s2n_offered_psk_list *identity_list)
 {
     ENSURE_REF(conn);
-    ENSURE_REF(identities);
+    ENSURE_REF(identity_list);
 
     struct s2n_array *known_psks = &conn->psk_params.psk_list;
     conn->psk_params.chosen_psk = NULL;
 
-    for (size_t i = 0; i < identities_length; i++) {
-        struct s2n_blob wire_identity = { 0 };
-        GUARD_AS_RESULT(s2n_blob_init(&wire_identity, identities[i].data, identities[i].length));
-
-        struct s2n_psk *local_match = NULL;
-        GUARD_RESULT(s2n_match_psk_identity(known_psks, &wire_identity, &local_match));
-
-        /* When a local match is found we do not end this loop early in an attempt
-         * to keep the server's known identities a secret and hide its ordering.
-         */
-        if (local_match != NULL && conn->psk_params.chosen_psk == NULL) {
-            conn->psk_params.chosen_psk_wire_index = i;
-            conn->psk_params.chosen_psk = local_match;
-        }
-    }
-
-    return S2N_RESULT_OK;
-}
-
-static S2N_RESULT s2n_count_psk_identities(struct s2n_stuffer *input, uint16_t *identity_count)
-{
-    ENSURE_REF(input);
-    ENSURE_REF(identity_count);
-
-    const size_t obfuscated_ticket_age_size = sizeof(uint32_t);
-
-    *identity_count = 0;
-    while (s2n_stuffer_data_available(input) > 0) {
-        uint16_t identity_size = 0;
-        GUARD_AS_RESULT(s2n_stuffer_read_uint16(input, &identity_size));
-        GUARD_AS_RESULT(s2n_stuffer_skip_read(input, identity_size));
-        GUARD_AS_RESULT(s2n_stuffer_skip_read(input, obfuscated_ticket_age_size));
-        (*identity_count)++;
+    for (size_t i = 0; i < known_psks->len; i++) {
```

Comment:
The change is intentional. We have an S2N tenet to prefer local order over wire order.

Once we support 0RTT we will have to make an exception for the first wire identity when attempting early data, but otherwise we should prefer the local order.